contract/services.py

```diff
@@ -351,6 +354,50 @@ def __copy_details(self, contract_id, amended_contract):
             cd_new.json_ext = None
             cd_new.save(username=self.user.username)
 
+    @check_authentication
+    def renew(self, contract):
+        try:
+            # check rights for delete contract
+            if not self.user.has_perms(ContractConfig.gql_mutation_renew_contract_perms):
+                raise PermissionError("Unauthorized")
+            from core import datetime, datetimedelta
+            contract_to_renew = ContractModel.objects.filter(id=contract["id"]).first()
+            contract_id = contract["id"]
+            # block deleting contract not in Updateable or Approvable state
+            state_right = self.__check_rights_by_status(contract_to_renew.state)
+            # check if we can amend
+            if state_right is not "cannot_update" and contract_to_renew.state is not ContractModel.STATE_TERMINATED:
+                raise ContractUpdateError("You cannot renew this contract!")
+            # create copy of the contract
+            renewed_contract = copy(contract_to_renew)
+            renewed_contract.id = None
+            # Date to (the previous contract) became date From of the new contract (TBC if we need to add 1 day)
+            # Date To of the new contract is calculated by DateFrom new contract + “Duration in month of previous contract“
+            length_contract = (contract_to_renew.date_valid_to.year - contract_to_renew.date_valid_from.year) * 12 \
+                              + (contract_to_renew.date_valid_to.month - contract_to_renew.date_valid_from.month)
+            renewed_contract.date_valid_from = contract_to_renew.date_valid_to + datetimedelta(days=1)
+            renewed_contract.date_valid_to = contract_to_renew.date_valid_to + datetimedelta(months=length_contract)
+            renewed_contract.state, renewed_contract.version = (ContractModel.STATE_DRAFT, 1)
+            renewed_contract.amount_rectified, renewed_contract.amount_due = (0, 0)
+            renewed_contract.save(username=self.user.username)
+            historical_record = renewed_contract.history.all().first()
+            renewed_contract.json_ext = json.dumps(_save_json_external(
+                user_id=historical_record.user_updated.id,
+                datetime=historical_record.date_updated,
+                message=f"contract renewed - state "
+                        f"{historical_record.state}"
+            ), cls=DjangoJSONEncoder)
+            renewed_contract.save(username=self.user.username)
+            # copy also contract details and contract contribution plan details
+            self.__copy_details(contract_id=contract_id, amended_contract=renewed_contract)
```

Comment:
@delcroip As I mentioned - here I copy the contract details based on previous contract details. I can make small update in that function to not call "amended_contract" so as to you were not confused when reading the code etc.